drivers/clock_control/clock_control_bl60x.c

```diff
@@ -0,0 +1,913 @@
+	int ret = -EINVAL;
+	uint32_t key;
+	enum bl60x_clkid oldroot;
+
+	key = irq_lock();
+
+	if ((enum bl60x_clkid)sys == bl60x_clkid_clk_crystal) {
+		if (data->crystal_enabled) {
+			ret = 0;
+		} else {
+			data->crystal_enabled = true;
+			ret = clock_control_bl60x_update_root(dev);
+			if (ret < 0) {
+				data->crystal_enabled = false;
+			}
+		}
+	} else if ((enum bl60x_clkid)sys == bl60x_clkid_clk_pll) {
+		if (data->pll_enabled) {
+			ret = 0;
+		} else {
+			data->pll_enabled = true;
+			ret = clock_control_bl60x_update_root(dev);
+			if (ret < 0) {
+				data->pll_enabled = false;
+			}
+		}
+	} else if ((int)sys == BFLB_FORCE_ROOT_RC32M) {
+		if (data->root.source == bl60x_clkid_clk_rc32m) {
+			ret = 0;
+		} else {
+			/* Cannot fail to set root to rc32m */
+			data->root.source = bl60x_clkid_clk_rc32m;
+			ret = clock_control_bl60x_update_root(dev);
+		}
+	} else if ((int)sys == BFLB_FORCE_ROOT_CRYSTAL) {
+		if (data->root.source == bl60x_clkid_clk_crystal) {
+			ret = 0;
+		} else {
+			oldroot = data->root.source;
+			data->root.source = bl60x_clkid_clk_crystal;
+			ret = clock_control_bl60x_update_root(dev);
+			if (ret < 0) {
+				data->root.source = oldroot;
+			}
+		}
+	} else if ((int)sys == BFLB_FORCE_ROOT_PLL) {
+		if (data->root.source == bl60x_clkid_clk_pll) {
+			ret = 0;
+		} else {
+			oldroot = data->root.source;
+			data->root.source = bl60x_clkid_clk_pll;
+			ret = clock_control_bl60x_update_root(dev);
+			if (ret < 0) {
+				data->root.source = oldroot;
+			}
+		}
+	}
+
+	irq_unlock(key);
+	return ret;
+}
+
+static int clock_control_bl60x_off(const struct device *dev, clock_control_subsys_t sys)
+{
+	struct clock_control_bl60x_data *data = dev->data;
+	int ret = -EINVAL;
+	uint32_t key;
+
+	key = irq_lock();
+
+	if ((enum bl60x_clkid)sys == bl60x_clkid_clk_crystal) {
+		if (!data->crystal_enabled) {
+			ret = 0;
+		} else {
+			data->crystal_enabled = false;
+			ret = clock_control_bl60x_update_root(dev);
+			if (ret < 0) {
+				data->crystal_enabled = true;
+			}
+		}
+	} else if ((enum bl60x_clkid)sys == bl60x_clkid_clk_pll) {
+		if (!data->pll_enabled) {
+			ret = 0;
+		} else {
+			data->pll_enabled = false;
+			ret = clock_control_bl60x_update_root(dev);
+			if (ret < 0) {
+				data->pll_enabled = true;
+			}
+		}
+	}
+
+	irq_unlock(key);
+	return ret;
+}
+
+static enum clock_control_status clock_control_bl60x_get_status(const struct device *dev,
+								clock_control_subsys_t sys)
+{
+	struct clock_control_bl60x_data *data = dev->data;
+
+	switch ((enum bl60x_clkid)sys) {
+	case bl60x_clkid_clk_root:
+	case bl60x_clkid_clk_bclk:
+	case bl60x_clkid_clk_rc32m:
+		return CLOCK_CONTROL_STATUS_ON;
+	case bl60x_clkid_clk_crystal:
+		if (data->crystal_enabled) {
+			return CLOCK_CONTROL_STATUS_ON;
+		}
+		return CLOCK_CONTROL_STATUS_OFF;
+	case bl60x_clkid_clk_pll:
+		if (data->pll_enabled) {
+			return CLOCK_CONTROL_STATUS_ON;
+		}
+		return CLOCK_CONTROL_STATUS_OFF;
+	default:
+		return -EINVAL;
```

Comment:
function returns a clock_control_status - I think you want `CLOCK_CONTROL_STATUS_UNKNOWN`?